pkg/syncer/specsyncer.go

```diff
@@ -77,58 +76,21 @@ func NewSpecSyncer(from, to *rest.Config, syncedResourceTypes []string, clusterI
 	}
 	fromClient := fromClients.Cluster(logicalClusterID)
 	toClient := dynamic.NewForConfigOrDie(to)
-	return New(fromDiscovery, fromClient, toClient, upsertIntoDownstream, deleteFromDownstream, func(c *Controller, gvr schema.GroupVersionResource) cache.ResourceEventHandlerFuncs {
-		return cache.ResourceEventHandlerFuncs{
-			AddFunc: func(obj interface{}) { c.AddToQueue(gvr, obj) },
-			UpdateFunc: func(oldObj, newObj interface{}) {
-				if !deepEqualApartFromStatus(oldObj, newObj) {
-					c.AddToQueue(gvr, newObj)
-				}
-			},
-			DeleteFunc: func(obj interface{}) { c.AddToQueue(gvr, obj) },
-		}
-	}, syncedResourceTypes, clusterID)
-}
-
-// TODO:
-// This function is there as a quick and dirty implementation of namespace creation.
-// In fact We should also be getting notifications about namespaces created upstream and be creating downstream equivalents.
-func (c *Controller) ensureNamespaceExists(namespace string) error {
-	namespaces := c.toClient.Resource(schema.GroupVersionResource{
-		Group:    "",
-		Version:  "v1",
-		Resource: "namespaces",
-	})
-	newNamespace := &unstructured.Unstructured{}
-	newNamespace.SetAPIVersion("v1")
-	newNamespace.SetKind("Namespace")
-	newNamespace.SetName(namespace)
-	if _, err := namespaces.Create(context.TODO(), newNamespace, metav1.CreateOptions{}); err != nil {
-		if !k8serrors.IsAlreadyExists(err) {
-			klog.Infof("Error while creating namespace %s: %v", namespace, err)
-			return err
-		}
-	}
-	return nil
+	return New(fromDiscovery, fromClient, toClient, KcpToPcluster, syncedResourceTypes, clusterID)
 }
 
 func deleteFromDownstream(c *Controller, ctx context.Context, gvr schema.GroupVersionResource, namespace, name string) error {
 	// TODO: get UID of just-deleted object and pass it as a precondition on this delete.
 	// This would avoid races where an object is deleted and another object with the same name is created immediately after.
 
+	// TODO: check to see if ns is deleting/deleted
```

Comment:
Is this something you need to resolve in this PR? What is the effect here if a namespace is being deleted?

Reply:
This comment is removed now, but I did file an issue around namespace deletion: https://github.com/kcp-dev/kcp/issues/398